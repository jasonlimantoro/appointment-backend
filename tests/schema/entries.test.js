import gql from 'graphql-tag';
import { createTestClientAndServer } from '../utils';
import mockedEntries from '../../fixtures/entries';
import mockedGuest from '../../fixtures/guests';

describe('Entry Schema', () => {
  it('listEntry: should work', async () => {
    const { query, entryAPI, guestAPI } = createTestClientAndServer();
    entryAPI.list = jest.fn().mockResolvedValue(mockedEntries);
    guestAPI.get = jest.fn().mockResolvedValue(mockedGuest[0]);
    const LIST_ENTRY = gql`
      query {
        listEntry {
          Guest {
            firstName
            lastName
            NIK
          }
          see
          createdAt
          endedAt
        }
      }
    `;
    const res = await query({ query: LIST_ENTRY });
    expect(entryAPI.list).toBeCalled();
    expect(guestAPI.get).toBeCalledTimes(mockedEntries.length);
    expect(res).toMatchSnapshot();
  });

  it('createEntry: should work', async () => {
    const { mutate, entryAPI, guestAPI } = createTestClientAndServer();
    const attributes = {
      id: 'asdf',
      see: 'xyz',
      Guest: {
        firstName: 'Jane',
        lastName: 'Doe',
        email: 'jane@example.com',
        company: 'Amazon',
        NIK: 'abcdefghijklmnop',
      },
    };
    const CREATE_ENTRY = gql`
      mutation CreateEntry($input: CreateEntryInput!) {
        createEntry(input: $input) {
          id
          see
          Guest {
            firstName
            lastName
            email
          }
        }
      }
    `;
    entryAPI.create = jest.fn().mockResolvedValue(attributes);
    guestAPI.findOrCreate = jest.fn().mockResolvedValue({ ...attributes.Guest, id: 'some-id' });
    guestAPI.get = jest.fn().mockResolvedValue(attributes.Guest);

    const res = await mutate({ mutation: CREATE_ENTRY, variables: { input: attributes } });

    expect(guestAPI.findOrCreate).toBeCalledWith(attributes.Guest);
    expect(entryAPI.create).toBeCalledWith({
      id: attributes.id,
      see: attributes.see,
      guestId: 'some-id',
    });

    expect(res).toMatchSnapshot();
  });

  it('getEntry: should work', async () => {
    const { query, entryAPI } = createTestClientAndServer();
    const GET_ENTRY = gql`
      query GetEntry($id: String!) {
        getEntry(id: $id) {
          id
          see
          createdAt
          endedAt
        }
      }
    `;
    const mock = mockedEntries[0];
    entryAPI.get = jest.fn().mockResolvedValue(mock);
    const res = await query({ query: GET_ENTRY, variables: { id: mock.id } });

    expect(entryAPI.get).toBeCalledWith(mock.id);

    expect(res).toMatchSnapshot();
  });

  it('endEntry: should work', async () => {
    const { mutate, entryAPI } = createTestClientAndServer();
    const END_ENTRY = gql`
      mutation EndEntry($id: String!) {
        endEntry(id: $id) {
          id
          see
          createdAt
          endedAt
        }
      }
    `;
    const mock = mockedEntries[0];
    entryAPI.end = jest
      .fn()
      .mockResolvedValue({ ...mock, endedAt: new Date(2020, 1, 1).toLocaleString() });
    const res = await mutate({ mutation: END_ENTRY, variables: { id: mock.id } });

    expect(entryAPI.end).toBeCalledWith(mock.id);
    expect(res).toMatchSnapshot();
  });
});
