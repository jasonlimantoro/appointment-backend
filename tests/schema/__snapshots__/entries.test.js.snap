// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entry Schema createEntry: should work 1`] = `
Object {
  "data": Object {
    "createEntry": Object {
      "Guest": Object {
        "email": "jane@example.com",
        "firstName": "Jane",
        "lastName": "Doe",
      },
      "id": "asdf",
      "see": "xyz",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Entry Schema endEntry: should work 1`] = `
Object {
  "data": Object {
    "endEntry": Object {
      "createdAt": "2018-09-10 04:05:33",
      "endedAt": "2020-02-01T00:00:00",
      "id": "123",
      "see": "person1",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Entry Schema getByGuestId: should work 1`] = `
Object {
  "data": Object {
    "byGuestId": Array [],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Entry Schema getEntry: should work 1`] = `
Object {
  "data": Object {
    "getEntry": Object {
      "createdAt": "2018-09-10 04:05:33",
      "endedAt": "2018-09-10 05:06:44",
      "id": "123",
      "see": "person1",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Entry Schema listEntry should be protected 1`] = `
Object {
  "data": Object {
    "listEntry": null,
  },
  "errors": Array [
    [GraphQLError: You are not authenticated],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Entry Schema listEntry: should work 1`] = `
Object {
  "data": Object {
    "listEntry": Array [
      Object {
        "Guest": Object {
          "NIK": "12312365",
          "firstName": "Andy",
          "lastName": "Huang",
        },
        "createdAt": "2018-09-10 04:05:33",
        "endedAt": "2018-09-10 05:06:44",
        "see": "person1",
      },
      Object {
        "Guest": Object {
          "NIK": "12312365",
          "firstName": "Andy",
          "lastName": "Huang",
        },
        "createdAt": "2018-02-07 16:00:04",
        "endedAt": null,
        "see": "person2",
      },
      Object {
        "Guest": Object {
          "NIK": "12312365",
          "firstName": "Andy",
          "lastName": "Huang",
        },
        "createdAt": "2019-06-21 17:00:45",
        "endedAt": "2019-06-21 18:00:00",
        "see": "person2",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Entry Schema listEntryToday: should work 1`] = `
Object {
  "data": Object {
    "listTodayEntry": Array [],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
